torch/distributed/_checkpoint/state_dict_loader.py

```diff
@@ -0,0 +1,96 @@
+from typing import Any, Callable, Dict, List
+import torch
+from torch.futures import Future
+from .metadata import ReadWriteRequest, Metadata
+from torch.distributed.shard.sharded_tensor import (
+    ShardedTensor,
+)
+from .storage_reader import StorageReader
+
+# -------------- private functions --------------
+def _reshard_and_prepare_read_request(state_dict: Dict[str, Any], metadata_from_stroage: Metadata) -> List[ReadWriteRequest]:
+    """
+    Use the loaded metadata and the current state dict to map the saved tensors to current tensors
+
+    NOTE:
+    The metadata loaded for storage is not reference in this code.
+    I did not implement ANY resharding at all. This code only work if the two model are exactly the same.
+    """
+    read_requests = []
+    for fqn, tensor in state_dict.items():
+        if isinstance(tensor, torch.Tensor):
+            # Reshape to 1D, the might make loading eaiser
+            target_tensor = tensor.detach().reshape(-1)
+            storage_size = target_tensor.nelement() * target_tensor.element_size()
+
+            rr = ReadWriteRequest(
+                target_tensor=target_tensor,
+                storage_key=fqn,
+                offset=0,
+                length=storage_size,
+            )
+
+            read_requests.append(rr)
+        elif isinstance(tensor, ShardedTensor):
+            for shard in tensor.local_shards():
+                suffix = "_".join([str(i) for i in shard.metadata.shard_offsets + shard.metadata.shard_sizes])
```

Comment:
Isn't `shard.metadata.shard_offsets` enough to generate a unique key? Why do we need to use `shard_sizes` as well?